File: src/TP3/Ej8/FriendCellPhoneBill.java
package TP3.Ej8;

public class FriendCellPhoneBill extends CellPhoneBill{
    private String[] friends;
    private int friendsIndex;
    private final int friendLimit;
    private double cost;
    private static final double DEFAULT_FRIEND_COST_PER_SECOND=0;
    FriendCellPhoneBill(String number, int limit){
        super(number);
        this.friendLimit=limit;
        this.friends=new String[limit];
        this.cost=DEFAULT_FRIEND_COST_PER_SECOND;
    }

    public void setCost(double cost) {
        this.cost = cost;
    }
    private boolean containsFriend(String number){
        for (String i:friends){
            if (i.equals(number)){
                return true;
            }
        }
        return false;
    }
    public void addFriend(String number){
        if(friendsIndex < friendLimit && !containsFriend(number)) {
            friends[friendsIndex++] = number;
        }
    }
    public void removeFriend(String number) {
        boolean found = false;
        for(int i = 0; i < friendsIndex && !found; i++) {
            if(friends[i].compareTo(number) == 0) {
                friends[i] = friends[--friendsIndex];
                found = true;
            }
        }
    }
    @Override
    public double processBill() {
        double total = 0;
        for (int i = 0; i < callsIndex; i++) {
            boolean friendlyCall = false;
            for(int j = 0; j < friendsIndex && !friendlyCall; j++) {
                if (calls[i].getTo().compareTo(friends[j]) == 0) {
                    friendlyCall = true;
                }
            }
            total += friendlyCall ? calls[i].getCost() * cost : calls[i].getCost();
        }
        return total;
    }


}

File: Ruby/Guias/G9/Ej4/friend_cell_phone_bill.rb
require_relative 'cell_phone_bill'
class FriendCellPhoneBill < CellPhoneBill
  DEFAULT_FRIEND_COST_PER_SECOND=0
  def initialize(my_number,limit)
    super my_number
    @limit = limit
    @numbers=[]
    @@friend_cost_per_second = DEFAULT_FRIEND_COST_PER_SECOND
  end
  def add_friend(number)
    raise "oiga uste eta aniadiendo a un numero ia aniadiho" if @numbers.include?number
    raise "nono papi te equivocaste" if @numbers.length == @limit
    @numbers.push(number)
  end
  def remove_friend(number)
    raise "caaaampeeeonn el numerito que me pasaste no esta en la lista de amigos, no flashees confianza" unless @numbers.include?(number)
    @numbers.delete(number)
  end
  def self.change_cost(cost)
    @@friend_cost_per_second = cost
  end

  def process_bill
    @calls.map { |c| @numbers.include?(c.to) ? @@friend_cost_per_second * c.cost : c.cost }.reduce(0, :+)
  end
end